src/workerd/api/filesystem.c++

```diff
@@ -0,0 +1,1213 @@
+    jsg::Lock& js) {
+  return js.alloc<ValueIterator>(IteratorState(JSG_THIS, collectEntries(js, inner)));
+}
+
+void FileSystemDirectoryHandle::forEach(jsg::Lock& js,
+    jsg::Function<void(
+        jsg::USVString, jsg::Ref<FileSystemHandle>, jsg::Ref<FileSystemDirectoryHandle>)> callback,
+    jsg::Optional<jsg::Value> thisArg) {
+  auto receiver = js.v8Undefined();
+  KJ_IF_SOME(arg, thisArg) {
+    auto handle = arg.getHandle(js);
+    if (!handle->IsNullOrUndefined()) {
+      receiver = handle;
+    }
+  }
+  callback.setReceiver(js.v8Ref(receiver));
+
+  for (auto& entry: collectEntries(js, inner)) {
+    callback(js, js.accountedUSVString(entry->getName(js)), entry.addRef(), JSG_THIS);
+  }
+}
+
+FileSystemFileHandle::FileSystemFileHandle(jsg::USVString name, kj::Rc<workerd::File> inner)
+    : FileSystemHandle(kj::mv(name)),
+      inner(kj::mv(inner)) {}
+
+jsg::Promise<bool> FileSystemFileHandle::isSameEntry(
+    jsg::Lock& js, jsg::Ref<FileSystemHandle> other) {
+  KJ_IF_SOME(other, kj::dynamicDowncastIfAvailable<FileSystemFileHandle>(*other)) {
+    return js.resolvedPromise(other.inner.get() == inner.get());
+  }
+  return js.resolvedPromise(false);
+}
+
+jsg::Promise<jsg::Ref<File>> FileSystemFileHandle::getFile(jsg::Lock& js) {
+  // TODO(node-fs): Currently this copies the file data into the new File object.
+  // Alternatively, File/Blob can be modified to allow it to be backed by a
+  // workerd::File such that it does not need to create a separate in-memory
+  // copy of the data. We can make that optimization as a follow-up, however.
+  auto stat = inner->stat(js);
+  return js.resolvedPromise(
+      js.alloc<File>(js, inner->readAllBytes(js), js.accountedUSVString(getName(js)), kj::String(),
+          (stat.lastModified - kj::UNIX_EPOCH) / kj::MILLISECONDS));
+}
+
+jsg::Promise<jsg::Ref<FileSystemWritableFileStream>> FileSystemFileHandle::createWritable(
+    jsg::Lock& js, jsg::Optional<FileSystemCreateWritableOptions> options) {
+  // TODO(node-fs): Per the spec, the stream is expected to write to a temporary
+  // file space that is only flushed to the underlying file when the stream is closed.
+  // This is also supposed to be paying attention to the `keepExistingData` option,
+  // which determines if the modified data retains the original file data or not.
+  // We are not yet implementing this and it needs to be done before we can ship this.
+  auto sharedState = kj::rc<FileSystemWritableFileStream::State>(inner.addRef());
+  auto stream =
+      js.alloc<FileSystemWritableFileStream>(newWritableStreamJsController(), sharedState.addRef());
+  stream->getController().setup(js,
+      UnderlyingSink{.type = kj::str("bytes"),
+        .write =
+            [state = sharedState.addRef()](
+                jsg::Lock& js, v8::Local<v8::Value> chunk, auto c) mutable {
+    // TODO(node-fs): The spec allows the write parameter to be a WriteParams struct
```

Comment:
This seems like a new syntax?

```suggestion
    // TODO(soon): The spec allows the write parameter to be a WriteParams struct
```

Reply:
I'm using `node-fs` intentionally to make it easier to find and track these specific todos